### PR TITLE
changed RestoreTransitionedObject API to use multipart if needed

### PR DESCRIPTION
**Pull request by poornas**
## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

### CODE REVIEW

File: cmd/erasure-object.go [hunk 1]
```diff
@@ -1385,34 +1385,160 @@ func (er erasureObjects) TransitionObject(ctx context.Context, bucket, object st
...
+	if len(actualfi.Parts) == len(fi.Parts) {
+		for i, pi := range actualfi.Parts {
+			if fi.Parts[i].Size != pi.Size {
+				partsMatch = false
+			}
+		}
+	}
+
+	if !partsMatch {
+		return setRestoreHeaderFn(oi, false, InvalidObjectState{Bucket: bucket})
```
Comment (krisis) [inline cmd/erasure-object.go]: Setting Object in InvalidObjectState would help in debugging.

File: cmd/erasure-object.go [hunk 2]
```diff
@@ -1385,34 +1385,160 @@ func (er erasureObjects) TransitionObject(ctx context.Context, bucket, object st
...
+	if rerr == nil {
+		oi.UserDefined[xhttp.AmzRestore] = fmt.Sprintf("ongoing-request=%t, expiry-date=%s", false, opts.Transition.RestoreExpiry.Format(http.TimeFormat))
+	} else { // allow retry in the case of failure to restore
+		oi.UserDefined[xhttp.AmzRestore] = fmt.Sprintf("ongoing-request=%t", false)
+	}
+	if _, err := er.CopyObject(ctx, bucket, object, bucket, object, oi, ObjectOptions{
+		VersionID: oi.VersionID,
+	}, ObjectOptions{
+		VersionID: oi.VersionID,
+		NoLock:    noLock, // lock already taken in restoreTransitionedObject
```
Comment (krisis) [inline cmd/erasure-object.go]: this comment is misleading. This method is called with noLock being `false` and `true` depending on where it's called.

File: cmd/erasure-object.go [hunk 3]
```diff
@@ -1385,34 +1385,160 @@ func (er erasureObjects) TransitionObject(ctx context.Context, bucket, object st
...
+	}); err != nil {
+		logger.LogIf(ctx, fmt.Errorf("Unable to update transition restore metadata for %s/%s(%s): %s", bucket, object, oi.VersionID, err))
 		return err
 	}
+	// cascade original error back after setting restore header
+	return rerr
```
Comment (krisis) [inline cmd/erasure-object.go]: I think the returning the original error the callers must be done outside of this function. The error return value should be used to indicate failure to update object metadata via CopyObject.

File: cmd/object-api-errors.go [hunk 4]
```diff
@@ -485,6 +485,12 @@ func (e TransitionStorageClassNotFound) Error() string {
 	return "Transition storage class not found "
 }
 
+type InvalidObjectState GenericError
+
+func (e InvalidObjectState) Error() string {
+	return "The operation is not valid for the current state of the object" + e.Bucket
```
Comment (krisis) [inline cmd/object-api-errors.go]: including e.Object would be useful.

File: cmd/erasure-object.go [hunk 5]
```diff
@@ -1385,34 +1385,160 @@ func (er erasureObjects) TransitionObject(ctx context.Context, bucket, object st
...
 	}
-	pReader := NewPutObjReader(hashReader)
-	ropts := putRestoreOpts(bucket, object, opts.Transition.RestoreRequest, oi)
-	ropts.UserDefined[xhttp.AmzRestore] = fmt.Sprintf("ongoing-request=%t, expiry-date=%s", false, opts.Transition.RestoreExpiry.Format(http.TimeFormat))
-	ropts.NoLock = true
-	if _, err := er.PutObject(ctx, bucket, object, pReader, ropts); err != nil {
-		return err
+
+	reducedErr := reduceWriteQuorumErrs(ctx, errs, objectOpIgnoredErrs, writeQuorum)
+	if reducedErr == errErasureWriteQuorum {
```
Comment (krisis) [inline cmd/erasure-object.go:1496]: Are we to ignore other non-nil values for reducedErr?

File: cmd/erasure-object.go [hunk 6]
```diff
@@ -1385,34 +1385,160 @@ func (er erasureObjects) TransitionObject(ctx context.Context, bucket, object st
...
+
+// restoreTransitionedObject for multipart object chunks the file stream from remote tier into the same number of parts
+// as in the xl.meta for this version and rehydrates the part.n into the fi.DataDir for this version as in the xl.meta
+func (er erasureObjects) restoreTransitionedObject(ctx context.Context, bucket string, object string, opts ObjectOptions) error {
+	defer func() {
+		ObjectPathUpdated(pathJoin(bucket, object))
+	}()
+	setRestoreHeaderFn := func(oi ObjectInfo, noLock bool, rerr error) error {
+		return er.updateRestoreMetadata(ctx, bucket, object, oi, opts, noLock, rerr)
+	}
```
Comment (krisis) [inline cmd/erasure-object.go]:
```suggestion
	setRestoreHeaderFn := func(oi ObjectInfo, noLock bool, rerr error) error {
		er.updateRestoreMetadata(ctx, bucket, object, oi, opts, noLock, rerr)
		return rerr
	}
```

File: cmd/erasure-object.go [hunk 7]
```diff
@@ -1385,34 +1385,160 @@ func (er erasureObjects) TransitionObject(ctx context.Context, bucket, object st
...
+	var oi ObjectInfo
+	// get the file info on disk for transitioned object
+	actualfi, _, _, err := er.getObjectFileInfo(ctx, bucket, object, opts, false)
+	if err != nil {
+		defer setRestoreHeaderFn(oi, false, toObjectErr(err, bucket, object))
```
Comment (krisis) [inline cmd/erasure-object.go]: why are we not returning here?

Comment (poornas) [reply cmd/erasure-object.go]: fixed

File: cmd/erasure-object.go (hunk 5 above)
Comment (poornas) [reply cmd/erasure-object.go:1496]: all we are doing is a rename at this point and as long as it is not writeQuorum, it should be ok. In any case, deferring to the CompleteMultipart implementation here for error handling

**Review by krishnasrinivas — APPROVED** [review]

**Pull request merged**